docs/networking/streamcodecs.md

```diff
@@ -0,0 +1,369 @@
+
+```java
+// Objects to create a stream codec for
+public record SimpleExample(String arg1, int arg2, boolean arg3) {}
+public record RegistryExample(double arg1, Holder<Item> arg2) {}
+
+// The stream codecs
+public static final StreamCodec<ByteBuf, SimpleExample> SIMPLE_STREAM_CODEC =
+    StreamCodec.composite(
+        // Stream codec and getter pair
+        ByteBufCodecs.STRING_UTF8, SimpleExample::arg1,
+        ByteBufCodecs.VAR_INT, SimpleExample::arg2,
+        ByteBufCodecs.BOOL, SimpleExample::arg3,
+        SimpleExample::new
+    );
+
+// Since this has a holder, a RegistryFriendlyByteBuf is used
+public static final StreamCodec<RegistryFriendlyByteBuf, RegistryExample> REGISTRY_STREAM_CODEC =
+    StreamCodec.composite(
+        // Note that ByteBuf stream codecs can be used here
+        ByteBufCodecs.DOUBLE, RegistryExample::arg1,
+        ByteBufCodecs.holderRegistry(Registries.ITEM), RegistryExample::arg2,
+        RegistryExample::new
+    );
+```
+
+### Transformers
+
+Stream codecs can be transformed into equivalent, or partially equivalent, representations using mapping methods. Two mapping methods apply to the value while one mapping method applies to the buffer.
+
+The `map` method transforms the value using two functions: one to transform the current type into the new type, and one to transform the new type back into the current type. This is analagous to [codec transformers][transformers].
+
+```java
+public static final StreamCodec<ByteBuf, ResourceLocation> STREAM_CODEC = 
+    ByteBufCodecs.STRING_UTF8.map(
+        // String -> ResourceLocation
+        ResourceLocation::new,
+        // ResourceLocation -> String
+        ResourceLocation::toString
+    );
+```
+
+The `apply` method transforms the value using a `StreamCodec.CodecOperation`. A `StreamCodec.CodecOperation` takes in a stream codec of the current type and returns a stream codec of the new type. These typically wrap around `map` or take in helper methods.
+
+```java
+public static final StreamCodec<ByteBuf, List<ResourceLocation>> STREAM_CODEC =
+    ResourceLocation.STREAM_CODEC.apply(ByteBufCodecs.list());
+```
+
+The `mapStream` method transforms the buffer using a function that takes in the new buffer type and returns the current buffer type. This method should rarely be used as most methods with stream codecs do not need to change the type of the buffer.
+
+```java
+public static final StreamCodec<RegistryFriendlyByteBuf, Integer> STREAM_CODEC =
+    ByteBufCodecs.VAR_INT.mapStream(buffer -> (ByteBuf) buffer);
+```
+
+### Unit
+
+A stream codec which supplies an in-code value and encodes to nothing can be represented using `StreamCodec#unit`. This is useful if no information should be synced across the network.
```

Comment:
It would probably be good to mention that the unit encoder does an equality check on the incoming value and throws if it's not equal to the value given when the unit codec is constructed. This means that non-record types without a valid `equals()` implementation require always passing the *exact* instance instead of a technically equivalent but newly constructed object.

Also, I don't like the example, an event bus seems like a *really* weird thing to put in a packet.